src/rust/engine/fs/src/snapshot.rs

```diff
@@ -0,0 +1,352 @@
+    file_digester: Arc<GFD>,
+    mut path_stats: Vec<PathStat>,
+  ) -> BoxFuture<Snapshot, String> {
+    let mut file_futures: Vec<BoxFuture<bazel_protos::remote_execution::FileNode, String>> =
+      Vec::new();
+    let mut dir_futures: Vec<BoxFuture<bazel_protos::remote_execution::DirectoryNode, String>> =
+      Vec::new();
+
+    for path_stat in &path_stats {
+      if path_stat.path().components().next().is_none() {
+        return futures::future::err(
+          "Snapshotting from path_stats: All PathStats must have at \
+least one path component, but at least one did not"
+            .to_string(),
+        ).to_boxed();
+      }
+    }
+
+    path_stats.sort_by(|a, b| a.path().cmp(b.path()));
+
+    for (first_component, group) in
+      &path_stats.clone().into_iter().group_by(|s| {
+        s.path().components().next().unwrap().as_os_str().to_owned()
+      })
+    {
+      let mut path_group: Vec<PathStat> = group.collect();
+      if path_group.len() == 1 && path_group.get(0).unwrap().path().components().count() == 1 {
+        // Exactly one entry with exactly one component indicates either a file in this
+        // directory, or an empty directory.
+        // If the child is a non-empty directory, or a file therein, there must be multiple
+        // PathStats with that prefix component, and we will handle that in the recursive
+        // from_path_stats call.
+
+        match path_group.pop().unwrap() {
+          PathStat::File { ref stat, .. } => {
+            let is_executable = stat.is_executable;
+            file_futures.push(
+              file_digester
+                .clone()
+                .digest(&stat)
+                .map_err(|e| format!("{:?}", e))
+                .and_then(move |digest| {
+                  let mut file_node = bazel_protos::remote_execution::FileNode::new();
+                  file_node.set_name(osstring_as_utf8(first_component)?);
+                  file_node.set_digest(digest.into());
+                  file_node.set_is_executable(is_executable);
+                  Ok(file_node)
+                })
+                .to_boxed(),
+            );
+          }
+          PathStat::Dir { .. } => {
+            // Because there are no children of this Dir, it must be empty.
+            dir_futures.push(
+              futures::done(
+                store
+                  .clone()
+                  .record_directory(&bazel_protos::remote_execution::Directory::new())
+                  .and_then(|digest| {
+                    let mut directory_node = bazel_protos::remote_execution::DirectoryNode::new();
+                    directory_node.set_name(osstring_as_utf8(first_component)?);
+                    directory_node.set_digest(digest.into());
+                    Ok(directory_node)
+                  }),
+              ).to_boxed(),
+            );
+          }
+        }
+      } else {
+        dir_futures.push(
+          // TODO: Memoize this in the graph
```

Comment:
This would be nice to do at some point, but I'm not going to rust to do it, because it's no worse than the current behaviour.

Reply:
Yea, this can wait.